File: src/main/scala/org/tindalos/principle/domain/agents/submodulesblueprint/OverlappingSubmoduleDefinitionsException.scala
package org.tindalos.principle.domain.agents.submodulesblueprint

import org.tindalos.principle.domain.expectations.exception.InvalidConfigurationException

class InvalidBlueprintDefinitionException(override val msg: String)
    extends InvalidConfigurationException(msg)

class OverlappingSubmoduleDefinitionsException(val overlaps: Set[Overlap])
  extends InvalidBlueprintDefinitionException(OverlappingSubmoduleDefinitionsException.toMessage(overlaps)) 

object OverlappingSubmoduleDefinitionsException {

  def toMessage(overlaps: Set[Overlap]) = {

    val msg = new StringBuffer("Overlapping submodules: ")
    overlaps.foreach({ overlap ⇒
      msg.append("\n")
      overlap.submoduleIds.foreach({ submoduleId ⇒ msg.append(submoduleId + " and ") })
      msg.append(msg.substring(0, msg.length() - 4))

    })

    msg.toString()
  }

}

File: src/main/java/org/tindalos/guardrails/internal/domain/constraints/submodules/OverlappingSubmoduleDefinitionsException.java
package org.tindalos.guardrails.internal.domain.constraints.submodules;

import java.util.Set;

/**
 * Exception thrown when submodule definitions overlap.
 * Overlapping submodules occur when the same package is defined in multiple submodules.
 */
public class OverlappingSubmoduleDefinitionsException extends InvalidBlueprintDefinitionException {

    private final Set<Overlap> overlaps;

    public OverlappingSubmoduleDefinitionsException(Set<Overlap> overlaps) {
        super(toMessage(overlaps));
        this.overlaps = overlaps;
    }

    public Set<Overlap> getOverlaps() {
        return overlaps;
    }

    private static String toMessage(Set<Overlap> overlaps) {
        StringBuilder msg = new StringBuilder("Overlapping submodules: ");

        for (Overlap overlap : overlaps) {
            msg.append("\n");
            Set<SubmoduleId> ids = overlap.submoduleIds();
            for (SubmoduleId submoduleId : ids) {
                msg.append(submoduleId).append(" and ");
            }
            // Remove the trailing " and " by going back 4 characters from the last append
            int currentLength = msg.length();
            if (currentLength > 0) {
                msg.delete(currentLength - 5, currentLength);
            }
        }

        return msg.toString();
    }

}

